Dihedral Plots: RDKit Mol Object

cadeduckworth opened the pull request:
- add RDKit Mol object to dihedral analysis plots
- add tests, and close #238 
- close #245

cadeduckworth commented:
@orbeckst 
I think I fixed the bond highlighting issue and image size issue.
I incorporated the new algorithm/methods with the 'low res' plotting functionality so that there is a functioning base to improve upon.

Currently going back and looking at our previous conversations and the new links you sent to improve upon the existing methods..

cadeduckworth commented:
![image](https://user-images.githubusercontent.com/104414304/220405639-05e05088-ce71-49f0-b89e-f8383f1e41c5.png)

orbeckst commented:
That looks better. Very good. Did you check other examples, too?

The other comments were along the lines
- add numerical labels (I shared a notebook with `prettyfy()` I think)
- add the dihedral name using atom numbers as well (... need to find original discussion)

cadeduckworth commented:
I checked the plots for that molecule that I have been testing on, but no others yet.

I changed a lot of things that will be helpful moving forward, especially for testing.
The information now used for Mol objects almost all come directly from `atom_indices`, which uses the RDKit conversion method that is tested and verified from PR#217.

For each dihedral group in a molecule the `atom_indices` and `bond_indices` are saved in a dictionary, where the key is in the format of `'C1-C2-C3-C4'` to correspond to MDAnalysis selection style, so that the results DataFrame can be the source that calls the associated plotting graphics info. That way, the DF determines which atom and bond indices are highlighted, rather than an iterative method.

cadeduckworth commented:
@orbeckst 
- found the notebook with `prettify()`, will add changes

cadeduckworth commented:
`atom_indices` added to label Mol object on plots

To-Do:
- fix image size to auto adjust by size of Mol object
- switch to SVG or enhance image quality

cadeduckworth commented:
![image](https://user-images.githubusercontent.com/104414304/220433031-d6220226-fd7b-40bf-9bcd-38f91f875263.png)

cadeduckworth commented:
cleaner plot titles

![image](https://user-images.githubusercontent.com/104414304/220435624-aebbb7dd-957c-469e-8b17-8df8e541dc6b.png)

cadeduckworth commented:
This is where I have landed with the combined RDKIT/matplotlib to SVG composition.
(SVG converted to PDF attached below)

With the number of atoms in the SAMPL9 molecules, and the small text size of the atom indices, the SVG plots will look much better for those purposes. It should also maintain a consistent box size for all of the molecules, so the orientation or overall size of the molecule should not cause the image to zoom in or out like the `MolToImage` RDKit PDF output method.

Regarding the PDF output method, I figured out that, because the default size of the `MolToImage` conversion is 300x300, 'square' shaped molecules end up appearing bigger under increased `extent` values, which means the 'scale' cannot really afford to be 'zoomed in' for horizontally oriented molecules that assume more of a 'linear' conformation. There exists a noticeable difference between the aspect options, `auto` and `equal`, but changing those without changing the frame size and other parameters causes distortion. All of this of course can be changed, but would be difficult to detect how to do so for each case. For the current methods that output PDF, the best combination I have come across is the most recently pushed commit (see also, plot directly beneath it), and likely suits most sizes/shapes (needs further testing to be sure).

Regarding the SVG output method, the `svgutil` code, for me, has been tricky for size, scale, and placement. There is a way to get the sizes after the original Mol or MPL objects are saved to SVG files which could make things easier, but it seems like minor changes in different steps of processing can cause big downstream changes for placement/proportions/scales. There is also a method that converts units between px, cm, pt, etc., but I have not had a chance to get into it yet.

It might be best to use what I have developed for SVGs as an option that can be specified, for now, and use the pdf method as default, unless conda has something similar to inkscape that we can include as a dependency in order to facilitate viewing, conversion, and notebook embedding?

[test.pdf](https://github.com/Becksteinlab/MDPOW/files/10798549/test.pdf)

orbeckst commented:
The PDF looks sharp and can be zoomed in ✅ 

The next thing is to make sure that we don't need to scale it to out it into a paper. To make it publication-ready we would want it to be "at final size", i.e., typical two-column wide figure (about 190 mm, just randomly used https://writing.stackexchange.com/questions/21658/what-is-the-image-size-in-scientific-paper-if-indicated-as-a-single-1-5-or-2-c ) and font size should not be smaller than 9 pt. I think the matplotlib default is 10 so if you make figures at final size, everything should look good.

orbeckst commented:
(see also https://github.com/Becksteinlab/making-prettier-plots/blob/master/Making%20better%20plots.ipynb )

cadeduckworth commented:
@orbeckst

Here are some plotting results after implementing existing method with `svgutils` and new method with `cairosvg` into `dihedrals.py` module.

`svgutils` is being used to create the SVG files/graphics and layer them.
`cairosvg` is being used to convert to pdf and scale/save in the desired size.
I added a kwarg to adjust the final exported pdf plot width, with the default set to 190mm.
I also added conversion to pixels for compatibility with `cairosvg.svg2pdf`.

Example with a fairly cluttered molecule from SAMPL9 set, before I added a little white space to the right border to allow for longer solvent names:
[S07_C21-C22-C25-F27_violins.pdf](https://github.com/Becksteinlab/MDPOW/files/10887920/S07_C21-C22-C25-F27_violins.pdf)

SM25 & SM26 after increasing whitespace on right border:
[SM25_C2-N3-S4-C5_violins.pdf](https://github.com/Becksteinlab/MDPOW/files/10887932/SM25_C2-N3-S4-C5_violins.pdf)
[SM26_O1-S2-C3-C4_violins.pdf](https://github.com/Becksteinlab/MDPOW/files/10887933/SM26_O1-S2-C3-C4_violins.pdf)

I have not been able to figure out how to suppress matplotlib output, which does not contain the Mol image.

Moving forward I am working on fixing the tests to accommodate what was changed during development. After I achieve full coverage again I will request a review and start making changes to the documentation to reflect changes and new functionality.

cadeduckworth commented:
@orbeckst 

The codecov report of 98.80% is misleading; some of my current tests are weak at best, and some are possibly not testing anything at all. Once the new SVG methods are finalized, more specific tests need to be written, existing tests need to be revisited, and fixture scopes need to be optimized.
Logger/error info needs to be added or updated. Once PR229 is merged the `test_workflows_base.py` module needs to test iterative usage of new plotting functionality for multiple projects. Docs will need a thorough rewrite. Tutorials and examples will need to be revisited.

The 'plotting only one solvent issue' needs to be revisited (see most recent comments, SM25 & SM26 pdfs). The current method almost works, but results in an unnecessary solvent label in the legend. I have already already come up with a way to probably fix this but have not tested it yet.

The plots look nice, they seem scalable (a few examples showed signs of adequate scaling, see also most recent comment for S07, SM255, SM26), and I am pleased with the combination of `svgutils` and `cairosvg`. I am not sure if the problem exists outside of Jupyter or IDEs, but I tried my hand at suppressing the `matplotlib` output with no luck.. and the output is a plot which does not contain the Mol object image, so it is annoying. I will look more into this.

cadeduckworth commented:
Reminder for myself:

https://github.com/Becksteinlab/making-prettier-plots/blob/master/Making%20better%20plots.ipynb

cadeduckworth commented:
@orbeckst 

For the plots, SVG to PDF no longer uses temporary files, and the one/two solvent plotting issue has been resolved. If everything looks good I will start cleaning things up and finishing the docs and tests.

orbeckst commented on mdpow/tests/test_automated_dihedral_analysis.py:
```diff
@@ -130,14 +136,15 @@ def test_build_universe(self, SM25_tmp_dir):
         solute = u.select_atoms('resname UNK')
         solute_names = solute.atoms.names
         assert solute_names.all() == self.universe_solute_atom_names.all()
+        #change this method
```
What is there TODO... then remove comment.

orbeckst replied:
still todo

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -31,13 +31,25 @@
 
 """
-
+#^need to update function list
```
remove/fix

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -31,13 +31,25 @@
 """
-
+#^need to update function list
 import os
```
`os.system` should not be needed, use `pathlib` or `os.path`.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -31,13 +31,25 @@
 from rdkit import Chem
+from rdkit.Chem import rdCoordGen
+
+# for SVG test
+#from rdkit.Chem.Draw import IPythonConsole
+#from IPython.display import SVG
```
remove

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -31,13 +31,25 @@
 import numpy as np
 import pandas as pd
 
+# add to dependencies for install and tests
```
remove

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -31,13 +31,25 @@
+# add to dependencies for install and tests
+import cairosvg
+import svgutils
+import svgutils.compose as sc
+import svgutils.transform as sg
```
Do you use `sc` and `sg` that often to make it worthwhile abbreviating it? The code is probably clearer when you just use `svgutils.transform`. 

Also, is `compose` still used? In any case, `svgutils.compose` is also clearer.

I'd prefer the full versions.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -157,18 +167,35 @@ def rdkit_conversion(u, resname):
-        u.add_TopologyAttr("elements", elements)
-        solute = u.select_atoms(f'resname {resname}')
-        mol = solute.convert_to('RDKIT')
+    # i think somewhere we decided to ditch try/except method, will check notes
+    # notes pertaining to testing topology attributes for explicit Hs
```
I don't remember. 

But document in the docstring that we _always_ guess elements using the MDA guesser.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -157,18 +167,35 @@ def rdkit_conversion(u, resname):
+    elements = [guess_atom_element(name) for name in u.atoms.names]
+    u.add_TopologyAttr("elements", elements)
+
+    # com or cog?
```
shouldn't matter

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -157,18 +167,35 @@ def rdkit_conversion(u, resname):
+
+    mol = solute.convert_to('RDKIT')
+    rdCoordGen.AddCoords(mol)
+    #rdDepictor.Compute2DCoords(mol), not sure how this is different from above^
```
remove if the above works

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -157,18 +167,35 @@ def rdkit_conversion(u, resname):
+    #rdDepictor.Compute2DCoords(mol), not sure how this is different from above^
+
+    for atom in mol.GetAtoms():
+        atom.SetProp("atomNote", str(atom.GetIdx()))
```
State in the docs what we are attaching as the label (is it the 0-based or 1-based atom index... I already forgot... so that's why docs are needed ;-) )

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -200,26 +227,43 @@ def dihedral_indices(dirname, resname, SMARTS=SMARTS_DEFAULT):
 
-    u = build_universe(dirname=dirname)
-    mol = rdkit_conversion(u=u, resname=resname)[0]
+    #u = build_universe(dirname=dirname)
+    #mol = rdkit_conversion(u=u, resname=resname)[0]
```
remove commented out code

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -252,15 +296,32 @@ def dihedral_groups(dirname, resname, SMARTS=SMARTS_DEFAULT):
-    atom_group_indices = dihedral_indices(dirname=dirname, resname=resname, SMARTS=SMARTS)
-    dihedral_groups = [solute.atoms[list(dihedral_indices)].names for dihedral_indices
-                       in atom_group_indices]
+    #u = build_universe(dirname=dirname)
+    #_, solute = rdkit_conversion(u=u, resname=resname)
+    #atom_indices = get_atom_indices(dirname=dirname, resname=resname, SMARTS=SMARTS)
```
cleanup

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -252,15 +296,32 @@ def dihedral_groups(dirname, resname, SMARTS=SMARTS_DEFAULT):
 
-def dihedral_groups_ensemble(dirname, atom_group_indices,
+def get_paired_indices(atom_indices, bond_indices, dihedral_groups):
+
```
add minimal doc string

orbeckst replied:
doc string still todo

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -252,15 +296,32 @@ def dihedral_groups(dirname, resname, SMARTS=SMARTS_DEFAULT):
+        dg_list.append(group)
+    all_dgs = tuple(dg_list)
+
+    if (len(atom_indices) == len(bond_indices) == len(all_dgs)):
```
remove outer parentheses (not needed)

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -462,38 +523,47 @@ def dihedral_violins(df, width=0.9, solvents=SOLVENTS_DEFAULT):
+    # to be plotted correctly
+    solvs = list(solvents)
+    if len(solvs) < 2:
+        solvs.append('N/A')
```
Can we use `None` instead of string `"N/A"`? That would be the canonical Python approach.

cadeduckworth replied:
Yes, if `None` will be displayed on the plot. Currently, two color bars exist in the legend regardless if one or two solvents are plotted, so `N/A` is displayed next to the unused color bar when only one solvent is plotted.

I am sure there is a way to remove the unused color bar from the legend when plotting only one solvent, I just haven't figured it out yet.

cadeduckworth replied:
@orbeckst reminder

orbeckst replied:
Can't we change the sns.catplot() to properly deal with the situation? See below...

orbeckst replied:
Btw, perhaps add an
```python
assert 0 < len(solvents) < 3
```
just to be sure (instead of a proper input check)

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -462,38 +523,47 @@ def dihedral_violins(df, width=0.9, solvents=SOLVENTS_DEFAULT):
                     kind="violin", split=True, width=width, inner=None, cut=0,
                     linewidth=0.5,
-                    hue_order=[solvs[0], solv2], col_order=["Coulomb", "VDW"],
+                    #hue_order=[solvs[0], solv2], col_order=["Coulomb", "VDW", "Structure"],
```
cleanup

orbeckst replied:
cleanup

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -462,38 +523,47 @@ def dihedral_violins(df, width=0.9, solvents=SOLVENTS_DEFAULT):
+    _ = axV.spines["left"].set_visible(False)
+
+    axIM = g.axes_dict['Structure']                     
+    #axIM.imshow(im, aspect='equal', origin='upper', extent=(-80.5, 379.5, -222.0, 222.0))
```
cleanup

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -462,38 +523,47 @@ def dihedral_violins(df, width=0.9, solvents=SOLVENTS_DEFAULT):
     axV = g.axes_dict['VDW']
-    axV.yaxis.set_visible(False)
-    axV.spines["left"].set_visible(False)
+    _ = axV.yaxis.set_visible(False)
+    _ = axV.spines["left"].set_visible(False)
```
I don't think that
```python
_ = ...
```
is necessary to suppress plotting.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -462,38 +523,47 @@ def dihedral_violins(df, width=0.9, solvents=SOLVENTS_DEFAULT):
-    axC.yaxis.set_major_locator(plt.matplotlib.ticker.MultipleLocator(60))
-    axC.yaxis.set_minor_locator(plt.matplotlib.ticker.MultipleLocator(30))
-    axC.yaxis.set_major_formatter(plt.matplotlib.ticker.FormatStrFormatter(r"$%g^\circ$"))
+    _ = axC.yaxis.set_major_locator(plt.matplotlib.ticker.MultipleLocator(60))
+    _ = axC.yaxis.set_minor_locator(plt.matplotlib.ticker.MultipleLocator(30))
+    _ = axC.yaxis.set_major_formatter(plt.matplotlib.ticker.FormatStrFormatter(r"$%g^\circ$"))
```
I don't think that
```python
_ = ...
```
is necessary to suppress plotting.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -462,38 +523,47 @@ def dihedral_violins(df, width=0.9, solvents=SOLVENTS_DEFAULT):
-    g.set_ylabels(r"dihedral angle $\phi$")
-    g.despine(offset=5)
+
+    _ = g.set_xlabels(r"$\lambda$")
+    _ = g.set_ylabels(r"dihedral angle $\phi$")
+    _ = g.despine(offset=5)
```
I don't think that
```python
_ = ...
```
is necessary to suppress plotting.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -462,38 +523,47 @@ def dihedral_violins(df, width=0.9, solvents=SOLVENTS_DEFAULT):
+                    hue_order=[solvs[0], solvs[1]], col_order=["Coulomb", "VDW", "Structure"],
                     sharex=False, sharey=True,
-                    height=2, aspect=2.5,
+                    height=3.0, aspect=2.0, # height was 3.0, aspect was 2.0
```
cleanup

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -462,38 +523,47 @@ def dihedral_violins(df, width=0.9, solvents=SOLVENTS_DEFAULT):
+
+    axIM = g.axes_dict['Structure']                     
+    #axIM.imshow(im, aspect='equal', origin='upper', extent=(-80.5, 379.5, -222.0, 222.0))
+    _ = axIM.axis('off')
```
I don't think that
```python
_ = ...
```
is necessary to suppress plotting.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -544,30 +614,65 @@ def plot_violins(df, resname, figdir=None, molname=None, width=0.9, solvents=SOL
     if molname is None:
         molname = resname
 
+        # figdir is now necessary??
```
cleanup or address

cadeduckworth replied:
I will remove the comment and leave the convention as is for now. `figdir` will not be necessary, but will be automatically generated once the 'project manifest' and 'results directory structure' changes are made.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -544,30 +614,65 @@ def plot_violins(df, resname, figdir=None, molname=None, width=0.9, solvents=SOL
+    plot_pdf_width = plot_pdf_width * 3.7795275591
+    # DEFAULT: 190 mm = 718.110236229 pixels
+
+    # put this into separate function
```
address?

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -544,30 +614,65 @@ def plot_violins(df, resname, figdir=None, molname=None, width=0.9, solvents=SOL
+    section = df.groupby(by="selection")
 
+    # conversion factor: 1 mm = 3.7795275591 px
+    plot_pdf_width = plot_pdf_width * 3.7795275591
```
Either
```python
    plot_pdf_width *= 3.7795275591
```
or (clearer)
```python
    plot_pdf_width_px = plot_pdf_width * 3.7795275591
```

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -544,30 +614,65 @@ def plot_violins(df, resname, figdir=None, molname=None, width=0.9, solvents=SOL
+
         plot = dihedral_violins(name[1], width=width, solvents=solvents)
-        plot.set_titles(f'{molname},{name[0]}, ''{col_name}')
+        _ = plot.set_titles(f'{molname}, {name[0]} {list(ab_pairs[name[0]][0])} | ''{col_name}')
```
Should work without the dummy assignment.

orbeckst replied:
I'd move `set_titles` to inside `dihedral_violins(..., title=f'{molname}, {name[0]} {list(ab_pairs[name[0]][0])} | ''{col_name}')` with a default of `None`. In this way, all plot styling is contained inside the function.

orbeckst replied:
And then you can also remove 4 lines of comment.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -544,30 +614,65 @@ def plot_violins(df, resname, figdir=None, molname=None, width=0.9, solvents=SOL
-    return plot
+        logger.info(f"Figure saved as {figfile}")
+
+    return logger.info(f"All figures generated and saved in {figdir}")
```
Returning the logger.info() does not make a lot of sense. 

```suggestion
    logger.info(f"All figures generated and saved in {figdir}")
    return
```

Right now, return `None` but then we'll update to return a df for all the created artifacts.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -544,30 +614,65 @@ def plot_violins(df, resname, figdir=None, molname=None, width=0.9, solvents=SOL
 def automated_dihedral_analysis(dirname=None, df_save_dir=None, figdir=None,
                                 resname=None, molname=None,
-                                SMARTS=SMARTS_DEFAULT,
+                                SMARTS=SMARTS_DEFAULT, plot_pdf_width=190,
```
document with units

orbeckst replied:
Still need entry for *plot_pdf_width* in doc string

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -683,4 +795,7 @@ def automated_dihedral_analysis(dirname=None, df_save_dir=None, figdir=None,
+    plot_violins(df_aug, resname=resname, molname=molname, mol=mol, plot_pdf_width=plot_pdf_width,
+                 ab_pairs=ab_pairs, figdir=figdir, width=width, solvents=solvents)
+
+    return logger.info(f"DihedralAnalysis completed for all projects in {dirname}")
```
Don't return the logger call.

orbeckst reviewed (changes requested):
Please see comments inline.

Update CHANGES under Enhancements.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -252,15 +281,29 @@ def dihedral_groups(dirname, resname, SMARTS=SMARTS_DEFAULT):
+    if len(atom_indices) == len(bond_indices) == len(all_dgs):
+        ab_pairs = {}
+        i = 0
+        while i < len(all_dgs):
+            ab_pairs[f'{all_dgs[i]}'] = (atom_indices[i], bond_indices[i])
+            i += 1
```
Can't we write this with a for loop
```python
for i in range(len(all_dgs)):
   ab_pairs[f'{all_dgs[i]}'] = (atom_indices[i], bond_indices[i])
```
Much cleaner than while.

orbeckst replied:
Actually, even better do a [dict comprehension](https://docs.python.org/3/tutorial/datastructures.html#dictionaries)
```python
ab_pairs = {all_dgs[i]: (atom_indices[i], bond_indices[i]) for i in range(len(all_dgs))}
```
Please change to make the code more contact (and faster, too, even though that's not a major concern here)

cadeduckworth replied:
I did not know you could do that, but immediately realized how it works compared to the current method, very cool!!

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -252,15 +281,29 @@ def dihedral_groups(dirname, resname, SMARTS=SMARTS_DEFAULT):
+        dg_list.append(group)
+    all_dgs = tuple(dg_list)
+
+    if len(atom_indices) == len(bond_indices) == len(all_dgs):
```
If this is just a sanity check and we assume it will always work (except for programming errors):
```python
assert len(atom_indices) == len(bond_indices) == len(all_dgs), "atom_indices and bond_indices are out of sync"
```
If there's a realistic chance that this could ACTUALLY fail (i.e., it depends on user input), then you need to raise an exception (and test the failure).

At the moment, the function will raise a NameError if the if statement is not true because you're returing `ab_pairs` without ecer defining them.

cadeduckworth replied:
the failure would be the result of program error, not user input, so I added the assert method as written above

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -252,15 +281,29 @@ def dihedral_groups(dirname, resname, SMARTS=SMARTS_DEFAULT):
     return dihedral_groups
 
-def dihedral_groups_ensemble(dirname, atom_group_indices,
+def get_paired_indices(atom_indices, bond_indices, dihedral_groups):
```
function needs a rewrite of the logic, see comments

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -252,15 +281,29 @@ def dihedral_groups(dirname, resname, SMARTS=SMARTS_DEFAULT):
-def dihedral_groups_ensemble(dirname, atom_group_indices,
+def get_paired_indices(atom_indices, bond_indices, dihedral_groups):
+
+    dg_list = []
+    for dg in dihedral_groups:
+        group = (f'{dg[0]}-{dg[1]}-{dg[2]}-{dg[3]}')
+        dg_list.append(group)
+    all_dgs = tuple(dg_list)
```
Why are you making the list a tuple?

Why not just build it with a list comprehension (... or tuple comprehension if you really want a tuple)?
```suggestion
    all_dgs = [f'{dg[0]}-{dg[1]}-{dg[2]}-{dg[3]}' for dg in dihedral_groups]
```

cadeduckworth replied:
I was using tuples during development so that it could not accidentally be modified downstream and because these values should not ever be changed externally, but your method is much cleaner so I will add the changes, thank you.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -252,15 +281,29 @@ def dihedral_groups(dirname, resname, SMARTS=SMARTS_DEFAULT):
+    all_dgs = tuple(dg_list)
+
+    if len(atom_indices) == len(bond_indices) == len(all_dgs):
+        ab_pairs = {}
+        i = 0
+        while i < len(all_dgs):
+            ab_pairs[f'{all_dgs[i]}'] = (atom_indices[i], bond_indices[i])
+            i += 1
```
More concise with dict comprehension and zip:
```suggestion
        ab_pairs = {group: (atoms, bonds) for group, atoms, bonds in zip(all_dgm atom_indices, bond_indices)}
```

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -491,9 +536,44 @@ def dihedral_violins(df, width=0.9, solvents=SOLVENTS_DEFAULT):
     return g
 
-def plot_violins(df, resname, figdir=None, molname=None, width=0.9, solvents=SOLVENTS_DEFAULT):
+def build_svg(mol, molname=None, name=None, ab_pairs=None,
```
Add a short doc string.

Any function that does not start with underscore is part of the public API and should be documented.

orbeckst reviewed (changes requested):
I did another review pass, please see comments. Thanks!

cadeduckworth commented:
@orbeckst 
All requested changes addressed (except for some docs) and new tests developed. Please do a quick review and if everything looks good, then I will do a final cleanup and complete all new documentation.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -549,25 +633,28 @@ def plot_violins(df, resname, figdir=None, molname=None, width=0.9, solvents=SOL
+    logger.info(f"All figures generated and saved in {figdir}")
 
-    return plot
+    return None
```
Returning the plot itself is generally a good idea as it allows for downstream processing. However, it looks as if these are multiple plots.

orbeckst replied:
(Nothing to be done here, will be part of #244 )

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -549,25 +633,28 @@ def plot_violins(df, resname, figdir=None, molname=None, width=0.9, solvents=SOL
+        figfile = pathlib.Path(newdir) / f"{molname}_{name[0]}_violins.pdf"
+        plot_pdf = cairosvg.svg2pdf(bytestring=fig.tostr(), write_to=str(figfile),
+                                    output_width=plot_pdf_width_px)
+
```
Might be worthwhile closing the fig here at the end to conserve resources
```
fig.close()
```
It also makes clearer that the figures are transient.

cadeduckworth replied:
can you close an `svgutils` figure object like that?

orbeckst replied:
I don't know — I *assumed* `fig` was a matplotlib Figure instance.

cadeduckworth replied:
@orbeckst

Can we use something like `fig.close()` somewhere to suppress the default `matplotlib` output (every MPL figure object created) when working in notebooks in a way that will not impact building the SVG figures from the MPL figure objects (which are not saved, but converted to SVG from memory)? I have not checked this in something like iPython, but in Jupyter you will get as output all of the figures before they are converted to SVG. This means the output just has a blank space where the RDKit Mol object is supposed to be. 

After the MPL figure is converted to SVG it is combined with the SVG render of the highlighted/labeled RDKit Mol object, then converted to and saved as a PDF file. Then, all of the PDF files are combined and additionally saved as one PDF file that contains all of the final figures for that dihedral atom group. I think the problem is that as matplotlib generates the figures it send them to output, but I do not know if closing them will remove them from memory and interfere with downstream processes. 

The incorrect output is not a pertinent issue, but it would be cool if there was a way to eliminate this without affecting subsequent processes. Instead of closing the figures, I believe there is a way to suppress the output, I just have not been able to figure it out yet.

orbeckst replied:
Do you mean something like https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.ioff.html ?

I am pretty sure that `fig.close()` removes the in-memory object. In fact, rendering a figure is a separate process from creating the figure object itself.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -666,21 +757,33 @@ def automated_dihedral_analysis(dirname=None, df_save_dir=None, figdir=None,
+    logger.info(f"DihedralAnalysis completed for all projects in {dirname}")
 
-    return plot_violins(df_aug, resname=resname, figdir=figdir, molname=molname, width=width, solvents=solvents)
+    return
```
With #244 will return interesting stuff...

orbeckst reviewed (changes requested):
Just some docs missing and minor suggestions.

EDIT: Also add entry to CHANGES!

cadeduckworth commented:
@orbeckst 
There is some documentation that is lacking or needs attention, but I am hesitant to commit to too much of a docs overhaul since much is anticipated to change in #244.

Is it possible that we can decide what should be minimally documented to avoid too many rewrites?

cadeduckworth commented:
@orbeckst 
I did a first pass on updating the docs and kwargs, but I will come back to it with fresh eyes later. I had to put a bandaid on the `figdir` situation (commit - [60636b4](https://github.com/Becksteinlab/MDPOW/pull/243/commits/60636b41bc99a6c2b49b0eeee0e79774ed84282f)) so the `base` module did not break, but as it is written, `figdir` really should be required as a positional argument for this PR. The current fix works pending issue #244.

Docs problems worth mentioning are intersphinx references for MDA `select_atoms` and the `svgutils` module.. haven't figured those out yet after including in `mdpow/doc/sphinx/source/conf.py`.. however the `svgutils` docs do not look like they support it, or I am looking in the wrong place. Also, I have not yet mentioned `cairosvg` so I need to test that one out in the docs too.

I am still having issues with tests for Python 3.7, so I will come back to this after the weekend.

I would like to finish up this PR as soon as possible and take care of #244 so we can move forward with SAMPL9.
If you get a chance to review this over the weekend that would be great, otherwise we can catch up on Monday.
Thanks, and have a great weekend!

cadeduckworth commented:
reminder for myself,
write test for the following:

`if molname is None:
    molname = resname`
    
```
test_resname = 'foo'
run_analysis(resname=test_resname)
assert molname == test_resname
```

orbeckst commented on CHANGES:
```diff
@@ -23,6 +23,9 @@ Changes
 
 Enhancements
 
+* convert figure components to SVG and save as scalable PDFs (#243)
+* add RDKit mol object to dihedrals plot with dihedral atom indices
+  labeled and dihedral atom group bonds highlighted (#243)
```
need to specify where — a user should be able to read CHANGES and figure out if anything affects them. You have to say in which module this and what the context is. Don't make people go to the issues.

orbeckst commented on mdpow/tests/test_automated_dihedral_analysis.py:
```diff
@@ -43,40 +46,79 @@ def SM25_tmp_dir(self, molname_workflows_directory):
+        mol, solute = dihedrals.rdkit_conversion(u=u, resname=resname)
+        return mol, solute
+
+    @pytest.fixture(scope="function")
```
function scope is the default — just remove when you use the default (here and elsewhere)

cadeduckworth replied:
There is an issue open for reviewing tests, so I will handle this here and leave a reminder in that issue to handle elsewhere.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -157,110 +180,151 @@ def rdkit_conversion(u, resname):
 
     """
 
-    try:
-        solute = u.select_atoms(f'resname {resname}')
-        mol = solute.convert_to('RDKIT')
-    except AttributeError:
-        elements = [guess_atom_element(name) for name in u.atoms.names]
-        u.add_TopologyAttr("elements", elements)
-        solute = u.select_atoms(f'resname {resname}')
-        mol = solute.convert_to('RDKIT')
+    elements = [guess_atom_element(name) for name in u.atoms.names]
+    u.add_TopologyAttr("elements", elements)
+
+    solute = u.select_atoms(f'resname {resname}')
+    solute.unwrap(compound='residues', reference='com')
+
+    mol = solute.convert_to('RDKIT')
```
Did you resort to always guessing because the elements are never present?

If so, write a comment stating this assumption and _change the docstring_ — it does not agree anymore with the code.

Otherwise the try/except is cleaner because it uses user data and only resorts to guessing when necessary.

cadeduckworth replied:
I cannot find the reason why it was removed, so I will revert to try/except and see how things go.

cadeduckworth replied:
It may have had something to do with our data that led me to believe that RDKit conversion always required this, rendering try/except useless. We should look into it more, but having the try/except present in the meantime hurts none at all.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -157,110 +180,151 @@ def rdkit_conversion(u, resname):
-def dihedral_indices(dirname, resname, SMARTS=SMARTS_DEFAULT):
-    '''Uses a SMARTS selection string to identify indices for
-       relevant dihedral atom groups.
+def get_atom_indices(mol, SMARTS=SMARTS_DEFAULT):
```
using `mol` is an improvement!! Much better to work with objects.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -254,38 +258,24 @@ def get_bond_indices(mol, atom_indices):
-           searches for .gro, .gro.bz2, .gro.gz, and .tpr files for topology,
-           and .xtc files for trajectory. It will default to using the tpr file
-           available.
+       *solute*
```
Just state that this is an MDAnalysis AtomGroup for the solute — be precise with the data type

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -299,6 +289,33 @@ def get_dihedral_groups(solute, atom_indices):
 def get_paired_indices(atom_indices, bond_indices, dihedral_groups):
+    '''Combines `atom_indices` and `bond_indices` in tuples
+       to be paired with their respective `dihedral_groups` in
+       :func:`~MDAnalysis.core.groups.select_atoms` selection string format.
```
I don't understand what `select_atoms()` has to do with this: Is this related to the selection *string* or the resulting AtomGroup?

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -299,6 +289,33 @@ def get_dihedral_groups(solute, atom_indices):
+       `atom_indices` and `bond_indices` are joined in a tuple
+       as the value, with the key being the respective member
+       of `dihedral_groups` to facilitate highlighting the 
+       relevant dihedral atom group when generating violin plots.
```
add an example to illustrate the abstract specification

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -299,6 +289,33 @@ def get_dihedral_groups(solute, atom_indices):
+       :returns:
+
+       *ab_pairs*
+           dictionary with key-value pair: '*#-*#-*#-*#': (atom_indices[i], bond_indices[i])
```
what does `*#` mean? `<ATOMNAME><ATOMNUMBER>` ?

cadeduckworth replied:
An example is C1-N2-O3-S4, so I changed it to list that as an example, since it is not technically the atom name or actual index number (the difference between 0 and 1-based atom index and label index is documented elsewhere).

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -490,13 +502,20 @@ def dihedral_violins(df, width=0.9, solvents=SOLVENTS_DEFAULT, plot_title=None):
+
+       *plot_title*
+           format: f'{molname}, {name[0]} {a} | ''{col_name}'
+           determined by,
```
Do you mean that the format string is determined by `build_svg()`? If so, make it part of the sentence and do not use seealso. If there's a hyperlink to follow as part of an explanatory sentence then that's good enough, no need for extra highlighting.

orbeckst replied:
seealso is typically use for broader context and related functionality where the connection may not be obvious from the other context

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -490,13 +502,20 @@ def dihedral_violins(df, width=0.9, solvents=SOLVENTS_DEFAULT, plot_title=None):
        :returns:
 
-       *violin plot*
+       *g (violin plot)*
```
just *g* as this cane be considered a variable name (but I am nit-picking here)

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -633,27 +691,36 @@ def plot_dihedral_violins(df, resname, mol, ab_pairs, figdir=None, molname=None,
 
-def automated_dihedral_analysis(dirname=None, df_save_dir=None, figdir=None,
-                                resname=None, molname=None,
+                                         #figdir=None is a temporary way to satisfy
+                                         #workflows base tests until issue #244 is resolved
+                                         #because it currently uses a **kwargs convention and the
+                                         #positional argument figdir will not carry over nicely, 
+                                         #or requires changes that will result in wasted time
+
+                                         #for this current version, figdir is required and will
+                                         #cause issues if not specified
```
indentation is odd — just put all the comments underneath the `def` — or put the `figdir=None` on a line of itself with the comment underneath in the middle of the arg list

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -781,6 +839,7 @@ def automated_dihedral_analysis(dirname=None, df_save_dir=None, figdir=None,
 
     df_aug = periodic_angle_padding(df, padding=padding)
 
+    #kwargs in weird order
```
Why? Can't you reorder?

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -759,7 +817,7 @@ def automated_dihedral_analysis(dirname=None, df_save_dir=None, figdir=None,
 
```
If you actually required `figdir` to be set (but will soon change to an positional arg anyway), add an assertion
```python
assert figdir is not None, "figdir MUST be set, even though it is a kwag. Will be changed with #244"
```

orbeckst reviewed (changes requested):
please see comments, mostly docs

orbeckst commented:
> mdpow/doc/sphinx/source/conf.py.. however the svgutils docs do not look like they support it, or I am looking in the wrong place.

See https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html for the intersphinx dict.

orbeckst commented:
Ping me if you have specific questions or if the PR is ready for review.

cadeduckworth commented:
reminder: check if intersphinx mapping works for `cairosvg`, `svgutils`, and `pypdf`

cadeduckworth commented:
@orbeckst It looks like some of the failed tests are related to RDKit. There may have been an update, so I will look into it.

cadeduckworth commented:
@orbeckst 
I think this might be the cause of four tests that fail starting with `Python>=3.9`: [RDKit issue](https://github.com/rdkit/rdkit/issues/5683)

The environment used for the COW analyses includes RDKit 2022.09.1, and the issue was addressed in a build of RDKit version 2023.03.1. The builds appear to be specific to Python versions, and it looks like some cases of index ordering (SMARTS) is the only relevant change causing those tests to fail.

`Python=3.7` CI uses the 2022 version, and `Python=3.8` uses a build from the 2023 version, but neither fail, so I am unsure how to solve this problem for all possible version combinations. At first glance, based on the atoms/bonds highlighted in our plots for COW analysis, it seems like there are no errors with the 2022 version (just different indices than some 2023 builds I guess).

orbeckst commented:
I looked at the test failures but I don't know what is the correct output. Is it possible that our current reference values are incorrect?

We could just require rdkit >=2023 (which is [available from 3.8 up](https://anaconda.org/conda-forge/rdkit/files)).

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -166,101 +189,146 @@ def rdkit_conversion(u, resname):
         solute = u.select_atoms(f'resname {resname}')
         mol = solute.convert_to('RDKIT')
 
+    solute.unwrap(compound='residues', reference='com')
```
For the code above: replace
```python
elements = [guess_atom_element(name) for name in u.atoms.names]
u.add_TopologyAttr("elements", elements)
```
with the more compact
```python
u.add_TopologyAttr("elements", guess_types(u.atoms.names))
```

cadeduckworth commented:
> I looked at the test failures but I don't know what is the correct output. Is it possible that our current reference values are incorrect?
> 
> We could just require rdkit >=2023 (which is [available from 3.8 up](https://anaconda.org/conda-forge/rdkit/files)) (and then axe 3.7).

@orbeckst The issue with this solution is that the 2023 build for 3.8 passes (the builds are version specific), with the same results as the 2022 build for 3.7. The change that causes test failure begins with RDKit 2023, Python >= 3.9.

cadeduckworth commented:
Another failed test unique to Python=3.7:

```
______________ TestAutomatedDihedralAnalysis.test_DataFrame_input ______________

self = <mdpow.tests.test_automated_dihedral_analysis.TestAutomatedDihedralAnalysis object at 0x7f854ad789d0>
SM25_tmp_dir = PosixPath('/tmp/pytest-of-runner/pytest-0/test_DataFrame_input0/SM25')
dihedral_data = (             selection solvent interaction lambda     time    dihedral
0         O1-S4-C6-C10   water     Coulomb   0...1000  50000.0  211.129909
15415  S4-O1-C5-C9   water         VDW   1000  50000.0  207.238198

[15416 rows x 6 columns])

    def test_DataFrame_input(self, SM25_tmp_dir, dihedral_data):
        df, _ = dihedral_data
        dihedrals.automated_dihedral_analysis(dirname=SM25_tmp_dir, figdir=SM25_tmp_dir,
                                              resname=resname, molname=molname,
                                              solvents=('water',), dataframe=df)
>       assert (SM25_tmp_dir / 'SM25' / 'SM25_C10-C5-S4-O11_violins.pdf').exists(), 'PDF file not generated'
E       AssertionError: PDF file not generated
E       assert False
E        +  where False = <bound method Path.exists of PosixPath('/tmp/pytest-of-runner/pytest-0/test_DataFrame_input0/SM25/SM25/SM25_C10-C5-S4-O11_violins.pdf')>()
E        +    where <bound method Path.exists of PosixPath('/tmp/pytest-of-runner/pytest-0/test_DataFrame_input0/SM25/SM25/SM25_C10-C5-S4-O11_violins.pdf')> = ((PosixPath('/tmp/pytest-of-runner/pytest-0/test_DataFrame_input0/SM25') / 'SM25') / 'SM25_C10-C5-S4-O11_violins.pdf').exists

mdpow/tests/test_automated_dihedral_analysis.py:310: AssertionError
```

orbeckst commented:
I am going to remove 3.7 #248 in a new PR, that's one thing less to worry about.

orbeckst commented:
@cadeduckworth , the MDAnalysis RDKit converter docs https://docs.mdanalysis.org/stable/documentation_pages/converters/RDKit.html#MDAnalysis.converters.RDKit.RDKitConverter state

> Changed in version 2.2.0: Improved the accuracy of the converter. Atoms in the resulting molecule now follow the same order as in the AtomGroup. 

That's relevant because the _Notes_ stated 
> The `_MDAnalysis_index` property of the resulting molecule corresponds to the index of the specific [AtomGroup](https://docs.mdanalysis.org/stable/documentation_pages/core/groups.html#MDAnalysis.core.groups.AtomGroup) that was converted, which may not always match the index property.

I _think_ this says that for MDA >= 2.2, we can assume that MDA atom indices and RDKIT mol indices agree.

orbeckst commented:
@cadeduckworth I merged #249 — please go ahead with your PR! 🚀

cadeduckworth commented:
> reminder for myself, write test for the following:
> 
> ```
> if molname is None:
>     molname = resname
> ```
> 
>  ```
> test_resname = 'foo'
> run_analysis(resname=test_resname)
> assert molname == test_resname
> ```

To-Do

cadeduckworth commented:
> reminder: check if intersphinx mapping works for `cairosvg`, `svgutils`, and `pypdf`

To-Do

cadeduckworth commented:
> @orbeckst
> 
> The codecov report of 98.80% is misleading; some of my current tests are weak at best, and some are possibly not testing anything at all. Once the new SVG methods are finalized, more specific tests need to be written, existing tests need to be revisited, and fixture scopes need to be optimized. Logger/error info needs to be added or updated. Once PR229 is merged the `test_workflows_base.py` module needs to test iterative usage of new plotting functionality for multiple projects. Docs will need a thorough rewrite. Tutorials and examples will need to be revisited.
> 
> The 'plotting only one solvent issue' needs to be revisited (see most recent comments, SM25 & SM26 pdfs). The current method almost works, but results in an unnecessary solvent label in the legend. I have already already come up with a way to probably fix this but have not tested it yet.
> 
> The plots look nice, they seem scalable (a few examples showed signs of adequate scaling, see also most recent comment for S07, SM255, SM26), and I am pleased with the combination of `svgutils` and `cairosvg`. I am not sure if the problem exists outside of Jupyter or IDEs, but I tried my hand at suppressing the `matplotlib` output with no luck.. and the output is a plot which does not contain the Mol object image, so it is annoying. I will look more into this.
> 
> EDIT: As a forewarning the changed files here are littered with comments and bad practices, so I will try to clean those up before you get a chance to review. This one required changes on more levels, so PR229 and PR242 are more complete and take precedence.

To-Do

cadeduckworth commented:
To-Do

- manually verify that all of the dihedral atom groups, atom indices, and bond indices are correct for both examples from the testing resources
- review documentation
- review code coverage and develop new tests as necessary 
- revise function and variable names if necessary

cadeduckworth commented on mdpow/workflows/dihedrals.py:
```diff
@@ -529,45 +678,66 @@ def plot_violins(df, resname, figdir=None, molname=None, width=0.9, solvents=SOL
-
     '''
 
+    assert figdir is not None, "figdir MUST be set, even though it is a kwarg. Will be changed with #244"
```
@orbeckst 

Please have a look at the modified test in `test_workflows_base.py` that handles this issue that will be corrected in the project manifest PR.

orbeckst replied:
That's ok, I just added a doc string as suggestion.

orbeckst commented on mdpow/workflows/base.py:
```diff
@@ -24,7 +24,7 @@
 import pandas as pd
 
-from mdpow.workflows import registry
+from ..workflows import registry
```
`base.py` is in the same directory as `registry.py`, isn't it?
```suggestion
from . import registry
```
so write it more simply!

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -426,30 +487,38 @@ def periodic_angle(df, padding=45):
            default: 0.9
 
        *solvents*
            The default solvents are documented under :data:`SOLVENTS_DEFAULT`.
```
document that this is normally a 2-tuple; state what happens when it only contains a single solvent

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -459,25 +528,24 @@ def dihedral_violins(df, width=0.9, solvents=SOLVENTS_DEFAULT):
-    solv2 = 'octanol'
-    if solvs.size > 1:
-        solv2 = solvs[1]
+    solvs = list(solvents)
```
I think we don't need to convert to a list if we use in catplot
```python
hue_order=solvents[:2], ...
```
If catplot _really_ needs a list, then we can use
```python
hue_order=list(solvents[:2]), ...
```

See below at catplot.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -459,25 +528,24 @@ def dihedral_violins(df, width=0.9, solvents=SOLVENTS_DEFAULT):
                     kind="violin", split=True, width=width, inner=None, cut=0,
                     linewidth=0.5,
-                    hue_order=[solvs[0], solv2], col_order=["Coulomb", "VDW"],
+                    hue_order=[solvs[0], solvs[1]], col_order=["Coulomb", "VDW", "Structure"],
```
If `solvents` only contains a single entry then we should have `hue_order=[solvents[0]]`. If it contains two entries we need `hue_order=solvents`. We can get this behavior by just slicing `[:2]`:
```suggestion
                    hue_order=solvents[:2], col_order=["Coulomb", "VDW", "Structure"],
```
Note that we don't need the conversion `solvs = list(solvents)` anymore.

Here's how slicing behaves
```python
In [1]: solvents = ("water", "octanol")

In [2]: solvents[:2]
Out[2]: ('water', 'octanol')

In [3]: solvents[:1]
Out[3]: ('water',)

In [4]: solvents[:1][:2]
Out[4]: ('water',)
```

cadeduckworth replied:
@orbeckst 
Can we manually set the available hues used for single solvent cases so that a plot containing only octane etc. is not blue? Not pertinent, but would be more consistent.. maybe leave alone for now and raise separate issue?

orbeckst replied:
probably, but as you said: not high priority, raising an issue would be sufficient

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -491,33 +559,114 @@ def dihedral_violins(df, width=0.9, solvents=SOLVENTS_DEFAULT):
+
        *figdir*
-           optional, path to the location to save figures
+           path to the location to save figures
```
```suggestion
           path to the location to save figures (REQUIRED but marked
           as a kwarg for technical reasons; will be changed in #244)
```

orbeckst reviewed (changes requested):
Getting there! Minor comments inline + whatever else you have on your TODO list.

orbeckst commented on mdpow/tests/test_automated_dihedral_analysis.py:
```diff
@@ -206,26 +265,22 @@ def test_save_fig_info(self, SM25_tmp_dir, caplog):
-        assert isinstance(plot, seaborn.axisgrid.FacetGrid)
+    def test_DataFrame_input(self, SM25_tmp_dir, dihedral_data):
+        df, _ = dihedral_data
+        dihedrals.automated_dihedral_analysis(dirname=SM25_tmp_dir, figdir=SM25_tmp_dir,
+                                              resname=resname, molname=molname,
+                                              solvents=('water',), dataframe=df)
```
This takes a while to run and is run in multiple tests as far as I can tell. Can we make it a module-level fixture and only run it once?

cadeduckworth replied:
We can do two fixtures for now with `scope=class` if you think it will reduce runtime. There are two distinct cases, each used twice, for four tests total. One case is the regular analysis and the other tests analysis with DataFrame input.

orbeckst replied:
cutting run time in half is already good, please do!

cadeduckworth replied:
@orbeckst my temporary testing directory uses `tmp_path` which is function scoped according to pytest. Since most other downstream tests call that fixture, they cannot exceed the function scope. Can we raise an issue to change to `tmp_dir_factory` in lieu of making these changes now?

cadeduckworth replied:
#235

orbeckst replied:
ok, add a comment to the current tests that they should be changed and reference #235

cadeduckworth commented:
@orbeckst 
Please see lines 143 and 144 (`dihedrals.build_universe`). If MDPOW supports single solvent Pow calculations (does it?), should we not always assume that results will be available for water by default?

cadeduckworth commented:
> reminder for myself, write test for the following:
> 
> ```
> if molname is None:
>     molname = resname
> ```
> 
>  ```
> test_resname = 'foo'
> run_analysis(resname=test_resname)
> assert molname == test_resname
> ```

To-Do

orbeckst commented:
It supports single solvent solvation free calculations, yes. You're right, the water directory does not necessarily exist.

However, just raise an issue that we improve `build_universe()` to be smarter about finding the topology. Right now it's working for everything that _we_ need.

Let's not use the ITP file because that might not be present, e.g., when someone else ran just the FEP on a cluster.

orbeckst commented:
Actually, your idea to use the first solvent seems pretty good and easy enough to implement. You could even make `solvent='water'` a kwarg and then the default is the current behavior.

cadeduckworth commented:
> It supports single solvent solvation free calculations, yes. You're right, the water directory does not necessarily exist.
> 
> However, just raise an issue that we improve `build_universe()` to be smarter about finding the topology. Right now it's working for everything that _we_ need.
> 
> Let's not use the ITP file because that might not be present, e.g., when someone else ran just the FEP on a cluster.

ITP file?

orbeckst commented on mdpow/workflows/base.py:
```diff
@@ -22,11 +22,11 @@
+import pandas as pd
 
-import logging
+from . import registry
```
see PR #257 — just merge my PR because it also fixes circular imports

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -13,47 +13,58 @@
+import svgutils.compose
+import svgutils.transform
 
 import mdpow
 from mdpow.analysis.dihedral import DihedralAnalysis
```
see PR #257

cadeduckworth replied:
@orbeckst I am looking over it now and then I will merge it, but I will have to manually merge the conflicts since my commit rearranged the inputs by category as outlined in PEP8.

cadeduckworth replied:
- whenever I merge `develop` into this PR

orbeckst replied:
sure, just merge develop into your PR and then manually resolve the conflicts

orbeckst replied:
once this PR is done, we can also consider to just run `black` on everything and forget about formatting

orbeckst commented on mdpow/tests/test_workflows_base.py:
```diff
@@ -62,17 +59,23 @@ def test_project_paths_csv_input(self, csv_input_data):
 
-        assert 'all analyses completed' in caplog.text, ('automated_dihedral_analysis '
-               'did not iteratively run to completion for the provided project')
+            assert 'all analyses completed' in caplog.text, ('automated_dihedral_analysis '
```
Unindent — that's better outside the pytest.raises block.

As a general rule reduce the amount of code in a with block to the essentials as this makes clearer what's relevant for testing the condition.

cadeduckworth replied:
I believe the test fails without the indent, possibly because of what is being tested or how it is written in the module. I am not certain of this, but making the change last night resulted in a passing test.

orbeckst commented:
> ITP file?

I was thinking along what I used for the dihedral sampler: use the initial files in the top level directories, including the topology file, but on second thought it makes no sense ebcause that files does not contain the solvent and is not useful for using with the trajectory. The TPR is really what we want.

Sorry to have confused you.

cadeduckworth commented:
[in response to](https://github.com/Becksteinlab/MDPOW/pull/243#discussion_r1248935370)

Docs say `hue_order` takes a list of strings, so we will do the list conversion as part of the kwarg.

cadeduckworth commented:
@orbeckst Ready for review, all requested changes resolved or responded to. I raised several new issues and commented or updated some existing issues to document what needs to be done soon, but is not immediately necessary for merging this PR. Minor reviewing left to do on my part tomorrow as a sanity check, but I hope to get this merged and resubmit the SAMPL9 jobs Monday afternoon.

orbeckst commented on mdpow/workflows/dihedrals.py:
```diff
@@ -531,19 +536,29 @@ def dihedral_violins(df, width=0.9, solvents=SOLVENTS_DEFAULT, plot_title=None):
+    # Usage in Jupyter causes matplotlib figure object output, not the completed figure
+    # Upcoming fix in issue #260
+    assert 0 < len(solvents) < 3, "one or two solvents must be specified, otherwise SOLVENTS_DEFAULT is used"
+    if len(list(solvents)) < 2: 
+        g = sns.catplot(data=df, x="lambda", y="dihedral", hue="solvent", col="interaction",
+                        kind="violin", split=False, width=width, inner=None, cut=0,
+                        linewidth=0.5,
+                        hue_order=list(solvents), col_order=["Coulomb", "VDW", "Structure"],
+                        sharex=False, sharey=True,
+                        height=3.0, aspect=2.0,
+                        facet_kws={'ylim': (-180, 180),
+                                'gridspec_kws': {'width_ratios': width_ratios,
+                                                    }})
+    else:
+        g = sns.catplot(data=df, x="lambda", y="dihedral", hue="solvent", col="interaction",
+                        kind="violin", split=True, width=width, inner=None, cut=0,
+                        linewidth=0.5,
+                        hue_order=list(solvents), col_order=["Coulomb", "VDW", "Structure"],
+                        sharex=False, sharey=True,
+                        height=3.0, aspect=2.0,
+                        facet_kws={'ylim': (-180, 180),
+                                'gridspec_kws': {'width_ratios': width_ratios,
+                                                    }})
```
That's a bit clunky and has a lot of repeated code. I suggest to unify (then you also don't have to test extra lines to make codecov happy... although in principle tests for 1 and 2 solvents _would_ be good)
```suggestion
    split = len(solvents) > 1
    g = sns.catplot(data=df, x="lambda", y="dihedral", hue="solvent", col="interaction",
                    kind="violin", split=split, width=width, inner=None, cut=0,
                    linewidth=0.5,
                    hue_order=list(solvents), col_order=["Coulomb", "VDW", "Structure"],
                    sharex=False, sharey=True,
                    height=3.0, aspect=2.0,
                    facet_kws={'ylim': (-180, 180),
                               'gridspec_kws': {'width_ratios': width_ratios}}
                    )

```
I think the two invocations only differed by `split` but if there are other things then I'd change the parameters and then feed the parameters to _one_ call to `catplot`.

cadeduckworth replied:
@orbeckst 
Hey Oliver, I made these changes. I also added a test, but our data in the testing resources only contain single solvent analyses and results, so it only tests the single solvent usage. I think that is okay for now, since the `assert` exists in the plotting code, and the only other option with two solvents is `split=True`.

orbeckst reviewed (changes requested):
Only one code compactifitransmogration.

orbeckst commented:
A momentous occasion: PR #243 has been merged. 🚀 🌟 

Congratulations!
